utils/reconnecting_udp_conn_test.go

```diff
@@ -0,0 +1,477 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	"github.com/uber/jaeger-client-go/log"
+)
+
+type mockResolver struct {
+	mock.Mock
+}
+
+func (m *mockResolver) ResolveUDPAddr(network string, hostPort string) (*net.UDPAddr, error) {
+	args := m.Called(network, hostPort)
+
+	a0 := args.Get(0)
+	if a0 == nil {
+		return (*net.UDPAddr)(nil), args.Error(1)
+	}
+	return a0.(*net.UDPAddr), args.Error(1)
+}
+
+type mockDialer struct {
+	mock.Mock
+}
+
+func (m *mockDialer) DialUDP(network string, laddr, raddr *net.UDPAddr) (*net.UDPConn, error) {
+	args := m.Called(network, laddr, raddr)
+
+	a0 := args.Get(0)
+	if a0 == nil {
+		return (*net.UDPConn)(nil), args.Error(1)
+	}
+
+	return a0.(*net.UDPConn), args.Error(1)
+}
+
+func newUDPListener() (net.PacketConn, error) {
+	return net.ListenPacket("udp", "127.0.0.1:0")
+}
+
+func newUDPListenerOnPort(port int) (net.PacketConn, error) {
+	return net.ListenPacket("udp", fmt.Sprintf("127.0.0.1:%d", port))
+}
+
+func newUDPConn() (net.PacketConn, *net.UDPConn, error) {
+	mockServer, err := newUDPListener()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", mockServer.LocalAddr().String())
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	conn, err := net.DialUDP("udp", nil, addr)
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	return mockServer, conn, nil
+}
+
+func assertSockBufferSize(t *testing.T, expectedBytes int, conn *net.UDPConn) bool {
+	fd, _ := conn.File()
+	bufferBytes, _ := syscall.GetsockoptInt(int(fd.Fd()), syscall.SOL_SOCKET, syscall.SO_SNDBUF)
+
+	// The linux kernel doubles SO_SNDBUF value (to allow space for bookkeeping overhead) when it is set using setsockopt(2), and this doubled value is returned by getsockopt(2)
+	// https://linux.die.net/man/7/socket
+	if runtime.GOOS == "linux" {
+		return assert.GreaterOrEqual(t, expectedBytes*2, bufferBytes)
+	}
+
+	return assert.Equal(t, expectedBytes, bufferBytes)
+}
+
+func assertConnWritable(t *testing.T, conn udpConn, serverConn net.PacketConn) (allSucceed bool) {
+	expectedString := "yo this is a test"
+	_, err := conn.Write([]byte(expectedString))
+	allSucceed = assert.NoError(t, err)
+
+	var buf = make([]byte, len(expectedString))
+	err = serverConn.SetReadDeadline(time.Now().Add(time.Second))
+	assertion := assert.NoError(t, err)
+	allSucceed = allSucceed && assertion
+
+	_, _, err = serverConn.ReadFrom(buf)
+	assertion = assert.NoError(t, err)
+	allSucceed = allSucceed && assertion
+	assertion = assert.Equal(t, []byte(expectedString), buf)
+
+	return allSucceed && assertion
```

Comment:
I am not seeing anything using the return value. Any reason not to use `require` instead of `assert` and fail the tests early? Logically the steps here must be done sequentially, no reason to continue if earlier steps fail.